### ISSUE TITLE
Serialize DoubleProperty of NaN

### ISSUE THREAD

**Issue by antoine-guillou** [joffrey-bion/fx-gson#7]
Using 
` Gson fxGson = FxGson.coreBuilder().serializeSpecialFloatingPointValues().create();`
does not allow to serialize DoubleProperty(Double.NaN) because we fall in `JsonWriter value(double value)` which is not built to take care of NaNs.

However, if I change the write method of DoublePropertyTypeAdapter to:
```
    @Override
    public void write(JsonWriter out, DoubleProperty value) throws IOException {
        Number nb = value.get();
        out.value(nb);
    }
```
it allows us to use `JsonWriter value(Number value)` which takes care of this special case.

I didn't test the deserialization and the performance impact yet, but I'd like to have your opinion on this change.

**Comment by joffrey-bion** [comment]
Very interesting, thanks for sharing!

This is clearly an issue, but I'm not really fond of this solution because I'm afraid we are bypassing the `serializeSpecialFloatingPointValues()` setting of Gson. We should not be taking care of the details of the serialization of the double ourselves.

In fact, this a global design problem in the library: we deal with the primitives ourselves, while we should let the Gson intercept and do what it pleases with this primitive. I would like to do an equivalent of what is done for `Property<T>` instead  (if not the exact same thing). Meaning, use a delegate adapter that would be chosen by the Gson.

This way, the current Gson config may deal with the double value as it is configured to. For instance, if you decide to write your own adaptation of `long` (which is a generalization of your current problem), we should use it when serializing long properties. This is currently what would be done for `Property<Long>` but not for `LongProperty`, unfortunately.

That would be the right thing to do but I'm not sure this is possible for primitive values. I'll have to check the Gson documentation when I'm home. Maybe we'll have to move to `Serializer`s instead of `TypeAdapter`s.

**Comment by joffrey-bion** [comment]
Ok, I have now fixed the root design problem.

Now, all my adapters (not only the collection ones and the object one) are properly delegating to Gson for the serialization of the inner values of the properties.

This way, all extra configuration and customization of Gson will be taken into account properly, including `serializeSpecialFloatingPointValues()`.

I'll try to release a new version ASAP containing this fix.

**Comment by antoine-guillou** [comment]
That's a neat solution 👍 
Thanks for the tool, very useful to me.
Cheers

**Comment by joffrey-bion** [comment]
Released in version 2.0.1
